beginner_source/ddp_fault_tolerance.rst

```diff
@@ -0,0 +1,212 @@
+mutable attribute of the training job necessary for its continuity.
+
+.. grid:: 2
+
+   .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
+      :shadow: none
+
+      -  Launching multi-GPU training jobs with ``torchrun``
+      -  Saving and loading snapshots of your training job
+      -  Structuring your training script for graceful restarts
+
+   .. grid-item-card:: :octicon:list-unordered;1em;` Prerequisites
+      :shadow: none
+
+      * High-level `overview <ddp_theory.html>`__ of DDP
+      * Familiarity with `DDP code <ddp_multigpu.html>`__ 
+      * A machine with multiple GPUs (this tutorial uses an AWS p3.8xlarge instance)
+      * PyTorch `installed <https://pytorch.org/get-started/locally/>`__ with CUDA
+      
+
+
+View the code used in this video: https://github.com/suraj813/distributed-pytorch/blob/main/multigpu_torchrun.py
+
+
+.. raw:: html
+
+   <div style="margin-top:10px; margin-bottom:10px;">
+     <iframe width="560" height="315" src="https://www.youtube.com/embed/9kIvQOiwYzg" frameborder="0" allow="accelerometer; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+   </div>
+
+
+Why use ``torchrun``
+~~~~~~~~~~~~~~~~~~~~
+
+``torchrun`` handles the minutiae of distributed training so that you
+don't need to. For instance,
+- You don't need to set environment
+variables or explicitly pass the ``rank`` and ``world_size``; torchrun
+assigns this along with several other `environment
+variables <https://pytorch.org/docs/stable/elastic/run.html#environment-variables>`__.
+- No need to call ``mp.spawn`` in your script; you only need a generic
+``main()`` entrypoint, and launch the script with ``torchrun``. This way
+the same script can be run in non-distributed as well as single-node and
+multinode setups. 
+- Gracefully restarting training from the last saved training
+snapshot
+
+
+Graceful restarts
+~~~~~~~~~~~~~~~~~~~~~
+For graceful restarts, it helps to have the following structure:
+
+.. code:: python
+
+   def main():
+     load_snapshot(snapshot_path)
+     initialize()
+     train()
+
+   def train():
+     for batch in iter(dataset):
+       train_step(batch)
+
+       if should_checkpoint:
+         save_snapshot(snapshot_path)
+
+If a failure occurs, torchrun will terminate all the processes and restart them. 
+Each process entrypoint first loads and initializes the last saved snapshot, and continues training from there.
+So at any failure, you only lose the training progress from the last saved snapshot. 
+
+In elastic training, whenever there are any membership changes (adding or removing nodes), torchrun will terminate and spawn processes
+on available devices. Having this structure ensures your training job can continue without manual intervention.
+
+
+
+
+
+Diff for `multigpu.py <https://github.com/suraj813/distributed-pytorch/blob/main/multigpu.py>`__ v/s `multigpu_torchrun.py <https://github.com/suraj813/distributed-pytorch/blob/main/multigpu_torchrun.py>`__
```

Comment:
Can you please do not reference your personal git repositories, i.e. add `multigpu.py` either to the same PR or submit it to `[pytorch/examples](https://github.com/pytorch/examples)`

Reply:
I'll submit to the Examples repo